---
title: "R Notebook"
output: html_notebook
---

```{r}
library(broom)
library(DESeq2)
library(tidyMB)
```

## Load in the data
In the data_wrangle.rmd script I put all of the count tables and metadata tables together.      
```{r}
ctData <- readRDS("~/At/Eugene/Analysis/data/countTable.rds")
ctMeta <- readRDS("~/At/Eugene/Analysis/data/metadata.rds")
peaks <- read_tsv("~/At/Eugene/Analysis/data/peaks.tsv")
```


## Making the DESeq Object
From my understanding, DESeq can directly test for interactions between experimental variables by comparing a reduced model to a full model. The downside to this is that the full model cannot directly test for genes that are affected by one variable and control for another. For example, with a model `~ Genotype + Trt + Genotype:Trt` and a reduced model of `~ Genotype + Trt` DESeq can find genes that show a genotype by treatment interaction. Unfortunately if I were to try to find genes that are affected by just Trt while controlling for Genotype and the Genotype x Treatment interaction, then it would only return the genes affected by treamtment in the first level of the Genotype factor. In this case, we would only be finding genes that are differentially expressed in the genotype Bur (since it is the first level of the Genotype factor). So, I'm making two models. I am interested in the genes showing a genotype x treatment interaction, but I am also interested in finding genes that are affected by genotype and treatment. One model can test for the interaction, and one can test for the genotype or trt effects.
```{r}
library(DESeq2)
dds <- DESeqDataSetFromMatrix(countData = ctTable, colData = ctMeta, design = ~ Genotype + Trt + Genotype:Trt)
dds_noInteraction <- DESeqDataSetFromMatrix(countData = ctTable, colData = ctMeta, design = ~Genotype + Trt)

keepers <- rowSums(counts(dds)) >= 30
dds <- dds[keepers,]
dds_noInteraction <- dds_noInteraction[keepers,]
dds_gxe <- DESeq(dds, test = "LRT", reduced = ~Genotype + Trt)
dds_noInteraction <- DESeq(dds_noInteraction)

```

```{r}
plotPCA(rld, intgroup = "Genotype", "Trt")
```


```{r}
gxe_res <- results(dds_gxe, pAdjustMethod = "bon")
gxe_genes <- gxe_res %>% 
  as.data.frame() %>% 
  rownames_to_column("Geneid") %>% 
  arrange(padj) %>% 
  filter(padj <= 0.05) 
```

```{r}
gxe_genes %>% 
  inner_join(peaks, by = "Geneid")
```


```{r}
vsdC <- assay(rld) %>% 
  as.data.frame() %>% 
  rownames_to_column("Geneid") %>% 
  gather(key = "SampleID", value = counts, -Geneid) %>% 
  inner_join(ctMeta, by = "SampleID")
```

```{r}
gxe_genes %>% 
  inner_join(peaks, by = "Geneid") %>% 
  inner_join(vsdC, by = "Geneid") %>% 
  group_by(Trt, Genotype, Geneid) %>% 
  summarise(meanV = mean(counts), se = sd(counts)/sqrt(n())) %>% 
  filter(Genotype %in% c("No", "Wu", "Hi")) %>% 
  ggplot(aes(Trt, meanV, group = Genotype, color = Genotype)) +
  geom_line() +
  geom_point(color = 'black') +
  facet_wrap(~Geneid, scales = "free", nrow = 1)
```


```{r}
mean_values_gxe <- vsdC %>% 
  inner_join(gxe_genes %>% select(Geneid, padj), by = "Geneid") %>% 
  group_by(Genotype, Trt, Geneid) %>% 
  summarise(meanv = mean(counts)) %>% 
  spread(Geneid, meanv)

mean_values_gxe <- as.data.frame(mean_values_gxe, 
                                row.names = paste(mean_values_gxe$Genotype, mean_values_gxe$Trt, sep = "_")) 

heatmap.2(as.matrix(mean_values_gxe[,3:ncol(mean_values_gxe)],), scale = "column", trace = "none")
```


```{r}
results(dds) %>% 
  as.data.frame() %>% 
  rownames_to_column("Geneid") %>% 
  na.omit() %>% 
  mutate(sig = ifelse(padj <= 0.001, "s", "ns")) %>% 
  mutate(color = ifelse(log2FoldChange > 0 & sig == "s", "up", 
                        ifelse(log2FoldChange < 0 & sig == "s", "down", "n"))) %>% 
  ggplot(aes(baseMean, log2FoldChange, color = color, alpha = sig)) +
  geom_point() +
  scale_color_manual(values = c("#440154FF", "grey50", "#21908CFF")) +
  scale_x_log10() +
  scale_alpha_manual(values = c(0.1, 0.5))

drought_genes <- results(dds) %>% 
  as.data.frame() %>% 
  rownames_to_column("Geneid") %>% 
  na.omit() %>% 
  filter(padj <= 0.001) %>% 
  mutate(type = ifelse(log2FoldChange > 0, "up", "down")) %>% 
  select(Geneid, type)
  
```

```{r}
rld <- rlog(dds)
```

```{r}
normV <- counts(dds, normalize = T) %>% 
  as.data.frame() %>% 
  rownames_to_column("Geneid") %>% 
  gather(key = "SampleID", value = "count", -Geneid) %>% 
  inner_join(ctMeta, by = "SampleID")



vsdC %>% 
  inner_join(drought_genes, by = "Geneid") %>% 
  group_by(Genotype, Trt, Geneid, type) %>% 
  summarise(meanV = mean(counts)) %>% 
  spread(Trt, meanV) %>% 
  mutate(log2FC = log2(T/C)) %>% 
  ggplot(aes(Genotype, log2FC, color = Genotype)) +
  geom_point(position = position_jitter(width = 0.1), alpha = 0.4) +
  geom_boxplot(color = 'black', outlier.size = 0) +
  facet_grid(.~type)

dPC <- plotPCA(rld, intgroup = c("Genotype", "Trt"), returnData = T, ntop = 500)
dPC %>%
  mutate(Treatment = ifelse(Trt == "T", "Drought", "Control")) %>% 
  ggplot(aes(PC1, PC2, color = Genotype, shape = Treatment)) +
  geom_point(size = 3) +
  theme_minimal() +
  guides(color = guide_legend(ncol = 2))
```

```{r}
vsdCTop <- vsdC %>% 
  group_by(Geneid) %>% 
  nest() %>% 
  mutate(variance = map_dbl(data, ~var(.x$counts))) %>% 
  arrange(-variance) %>% 
  head(1000) %>% 
  unnest()

gene_anova <- vsdCTop %>% 
  group_by(Geneid) %>% 
  nest() %>% 
  mutate(mod = map(data, ~aov(counts ~ Genotype*Trt, .))) %>% 
  unnest(map(mod, ~tidy(.))) %>% 
  na.omit() %>% 
  mutate(p.adj = p.adjust(p.adjust(p.value)))

gene_anova %>% 
  filter(term == "Genotype:Trt") %>% 
  filter(p.adj <= 0.01) %>% 
  arrange(-statistic)
```

```{r}
vsdC %>% 
  filter(Geneid == "AT1G58270") %>% 
  ggplot(aes(Genotype, counts, color = Trt)) +
  geom_boxplot()
```



```{r}
cdds <- DESeqDataSetFromMatrix(countData = colCounts, colData = colMeta, design = ~Trt)
cdds <- DESeq(cdds)
cdds.res <- results(cdds)
cdds.res %>% 
  as.data.frame() %>% 
  rownames_to_column("Geneid") %>% 
  as_tibble() %>% 
  mutate(sig = ifelse(padj <= 0.01, "s", "ns")) %>% 
  ggplot(aes(baseMean, log2FoldChange, color = sig, alpha = sig)) +
  geom_point() +
  scale_color_manual(values = c("black", "red")) +
  scale_x_log10() +
  scale_alpha_manual(values = c(0.2, 0.8))

colDrGenes <- cdds.res %>% 
  as.data.frame() %>% 
  rownames_to_column("Geneid") %>% 
  as_tibble() %>% 
  filter(padj <= 0.01 & log2FoldChange > 0) %>%
  mutate(type = "ColDrUp") %>% 
  select(Geneid, type)

colNorm <- counts(cdds, normalized = T) %>% 
  as.data.frame() %>% 
  rownames_to_column("Geneid") %>% 
  gather(key = "SampleID", value = "count", -Geneid) %>% 
  inner_join(colMeta, by = "SampleID")

normV %>% 
  inner_join(TvsC_drought) %>% 
  group_by(Genotype, Trt, Geneid, type) %>% 
  summarise(meanV = mean(count)) %>% 
  spread(Trt, meanV) %>% 
  mutate(log2FC = log2(T/C)) %>% 
  #group_by(Genotype) %>% 
  #nest() %>% 
  #mutate(meanv = map_dbl(data, ~mean(.x$log2FC[is.finite(.x$log2FC)], finite = T))) %>% 
  #arrange(meanv) %>% 
  #mutate(order = 1:n()) %>% 
  #unnest() %>% 
  #ungroup() %>% 
  ggplot(aes(Genotype, log2FC, color = Genotype)) +
  geom_boxplot() +
  geom_point(position = position_jitter(width = 0.2), alpha = 0.5) +
  facet_grid(.~type)
plotPCA(dds)
```